File: _314_BinaryTreeVerticalOrderTraversal/_314_BinaryTreeVerticalOrderTraversal.py
"""
本代码来自 CSON 官方视频课程，由 @CSON 提供
题目来源：http://leetcode.com
- CSON 致力于在 CS 领域内帮助中国人找到工作，让更多海内外国人受益
- 讲师：Edward Shi
- 官方网站：https://cspiration.com
- 版权归 CSON 所有，转发请注明出处
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    low = 0
    high = 0

    def verticalOrder(self, root: TreeNode) -> List[List[int]]:
        result = []
        if root == None: return result
        self.helper(root, 0)
        result = [[] for _ in range(self.low, self.high + 1)]
        queue = []
        index = []
        index.append(-self.low)
        queue.append(root)
        while len(queue) != 0:
            cur = queue.pop(0)
            idx = index.pop(0)
            result[idx].append(cur.val)
            if cur.left != None:
                queue.append(cur.left)
                index.append(idx - 1)
            if cur.right != None:
                queue.append(cur.right)
                index.append(idx + 1)
        return result

    def helper(self, root: TreeNode, idx: int) -> None:
        if root == None: return
        self.low = min(self.low, idx)
        self.high = max(self.high, idx)
        self.helper(root.left, idx - 1)
        self.helper(root.right, idx + 1)

File: 314-binary-tree-vertical-order-traversal/314-binary-tree-vertical-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> lst = new ArrayList<>();
        
        if(root == null){
            return lst;
        }
        
        Map<Integer, List<Integer>> map = new HashMap<>();
        Queue<TreeNode> q = new LinkedList<>();
        Queue<Integer> cols = new LinkedList<>();
        
        q.offer(root);
        cols.add(0);
        
        int max = 0;
        int min = 0;
        
        while(!q.isEmpty()){
            TreeNode cur = q.poll();
            int col = cols.poll();
            
            if(!map.containsKey(col)){
                map.put(col, new ArrayList<>());
            }
            
            map.get(col).add(cur.val);
            
            if(cur.left != null){
                q.offer(cur.left);
                cols.add(col - 1);
                
                min = Math.min(min, col - 1);
            }
            
            if(cur.right != null){
                q.offer(cur.right);
                
                cols.add(col + 1);
                
                max = Math.max(max, col + 1);
            }
        }
        
        for(int i = min; i <= max; i++){
            lst.add(map.get(i));
        }
        
        return lst;
    }
}